nbgitpuller/hookspecs.py

```diff
@@ -0,0 +1,21 @@
+import pluggy
+
+hookspec = pluggy.HookspecMarker("nbgitpuller")
+
+
+@hookspec
+def handle_files(self, repo, repo_parent_dir):
```

Comment:
I think this should instead take *all* the query params passed to nbgitpuller, and a 'target directory' (see review comment https://github.com/jupyterhub/nbgitpuller/pull/194/files#diff-d4ed3eb90f2e3f1b059692aeefaa13a8c67db3016cc824df0573b60f05d154cbR6), and `yield` progress lines to be shown to the user. This is how the current code shows the output of running commands to the terminal in nbgitpuller (see  https://github.com/jupyterhub/nbgitpuller/blob/4d40c2d0445213e1b94a63e2940cccc841f1ced3/nbgitpuller/pull.py#L254), and is very helpful for the user to see what exactly is happening right now, and to copy error messages when things fail.

It'll also need to be *async* - otherwise the whole web server will 'block' while the download is happening. If you look at https://github.com/jupyterhub/nbgitpuller/blob/4d40c2d0445213e1b94a63e2940cccc841f1ced3/nbgitpuller/pull.py#L12, you can see that we call git asynchronously (with a `Popen` rather than `check_output`), and `yield` each line of output as progress. We also `yield` the name of the command we are executing so the user can see that too. This function should thus adopt this pattern as well, and become an async generator

Reply:
We can use exceptions to signal failure, and need not use return codes for anything.